packages/swapper/src/swappers/RelaySwapper/utils/getTrade.ts

```diff
@@ -0,0 +1,365 @@
+        const appFeesAsset = quote.fees?.app?.currency
+          ? relayTokenToAsset(quote.fees.app.currency, deps.assetsById)
+          : undefined
+
+        // @TODO: we might need to change this logic when solana and BTC are supported
+        const isNativeCurrencyInput =
+          isNativeEvmAsset(sellAsset.assetId) && sellAsset.chainId === appFeesAsset?.chainId
+
+        // For cross-chain: always add back app fees
+        // For same-chain: only add back if input is native currency
+        if (isCrossChain || isNativeCurrencyInput) {
+          return quote.fees.app.amount
+        }
+      }
+
+      // cross-chain or same-chain with native currency as input are not applicable
+      return '0'
+    })()
+
+    const relayerFeesBuyAssetBaseUnit = (() => {
+      const relayerFeeRelayToken = quote.fees?.relayer?.currency
+      if (relayerFeeRelayToken && isValidRelayToken(relayerFeeRelayToken)) {
+        const relayerFeesAsset = relayTokenToAsset(relayerFeeRelayToken, deps.assetsById)
+        const relayerFeeAmount = quote.fees?.relayer?.amount ?? '0'
+
+        // If fee is already in buy asset, return as is
+        if (relayerFeesAsset?.assetId === buyAsset.assetId) {
+          return relayerFeeAmount
+        }
+
+        // if fee is in sell asset, convert to buy asset
+        if (relayerFeesAsset?.assetId === sellAsset.assetId) {
+          return convertPrecision({
+            value: relayerFeeAmount,
+            inputExponent: relayerFeesAsset.precision,
+            outputExponent: buyAsset.precision,
+          })
+            .times(rate)
+            .toFixed(0)
+        }
+
+        // If fee is in a different asset, convert to buy asset
+        const feeAmountUsd = quote.fees?.relayer?.amountUsd ?? '0'
+        const buyAssetUsd = quote.details?.currencyOut?.amountUsd ?? '0'
+        const buyAssetAmountBaseUnit = quote.details?.currencyOut?.amount ?? '1'
+
+        if (feeAmountUsd && buyAssetUsd && buyAssetAmountBaseUnit) {
+          // Calculate the rate: (buyAssetAmount / buyAssetUsd) gives us "buy asset per USD"
+          // Then multiply by feeAmountUsd to get the equivalent buy asset amount
+          const buyAssetCryptoBaseUnitPerUsd = bnOrZero(buyAssetAmountBaseUnit).div(buyAssetUsd)
+          const buyAssetFeesCryptoBaseUnit = bnOrZero(feeAmountUsd).times(
+            buyAssetCryptoBaseUnitPerUsd,
+          )
+
+          return buyAssetFeesCryptoBaseUnit.toFixed(0)
+        }
+
+        return '0'
+      }
+
+      return '0'
+    })()
+
+    // Add back relayer service and gas fees (relayer is including both) since they are downsides
+    // And add appFees
+    const buyAmountBeforeFeesCryptoBaseUnit = bnOrZero(currencyOut.minimumAmount)
+      // @blocking: relayer can be ETH or destination token
+      .plus(relayerFeesBuyAssetBaseUnit)
+      .plus(appFeesBaseUnit)
+      .toFixed()
+
+    return {
+      allowanceContract: hasApprovalStep ? swapSteps[0]?.items?.[0]?.data?.to : undefined,
```

Comment:
`data` is still typed as any here

Reply:
We've now lost type safety on `selectedItem.data` `selecteItem.data[prop]` accessing.
Is this correct? Are we confident those fields will never throw runtime errors when being accessed?

<img width="731" alt="image" src="https://github.com/user-attachments/assets/2d02a070-1066-47dc-889e-fae675b03260" />
<img width="638" alt="image" src="https://github.com/user-attachments/assets/f820c79d-7407-4d7b-9d54-44ac5ea6e350" />

Reply:
No, we didn't loose type safety because of [61c1ee7](https://github.com/shapeshift/web/pull/9188/commits/61c1ee796b2b16b960897c673e0a14257ca5302a) that made them optional as they might be not present because of EVM or SVM or UTXO

And then before getting the unsigned transaction we have a safe guard on them to avoid launching TXs without data which would be a big issue considering it would lock user funds

Reply:
I mean... We did on the revision I was testing 😉 
Looks gucci now with all partials

<img width="653" alt="image" src="https://github.com/user-attachments/assets/2c711bcc-2804-4801-97b4-eaec48f2085b" />